dali/image/tiff.h

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMAGE_TIFF_H_
+#define DALI_IMAGE_TIFF_H_
+
+#include <opencv2/opencv.hpp>
+
+#include <string>
+
+#include <vector>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/tensor.h"
+
+namespace dali {
+
+
+/**
+ * Class, that handles byte buffer for tiff image
+ */
+class tiff_buffer {
+ public:
+    explicit tiff_buffer(const std::string &buffer, bool little_endian = false) :
+            little_endian_(little_endian) {
+        stream_ = std::istringstream(buffer);
+        buffer_size_ = buffer.length();
+    }
+
+
+    /**
+     * Reads value of given type (byte length), at given offset in the buffer.
+     * This function guarantees, that such value can be read.
+     * @tparam ValueType Type of return value
+     * @param offset Offset (in bytes) from beginning of the buffer
+     * @return value
+     */
+    template<typename ValueType>
+    ValueType read(unsigned int offset = 0) {
+        assert(stream_.good());
+        assert(offset + sizeof(ValueType) < buffer_size_);
+        static_assert(std::is_integral<ValueType>::value, "Only integral values supported");
+
+        stream_.seekg(offset);
+        ValueType ret;
+        stream_.read(reinterpret_cast<char *>(&ret), sizeof(ValueType));
+        if (little_endian_) {
+            convert_le(&ret);
+        }
+        assert(stream_.good());
+        return ret;
+    }
+
+
+ private:
+    /**
+     * Converts given value in big-endian representation to little-endian
+     * @tparam T
+     * @param value pointer to given value
+     */
+    template<typename T>
+    void convert_le(T *value) {
```

Comment:
I would propose a different name to_little_endian. It's more expressive

Reply:
Good point